crates/proto/src/clients/mod.rs

```diff
@@ -0,0 +1,348 @@
+}
+
+impl MetadataInterceptor {
+    /// Adds or overwrites HTTP ACCEPT metadata to the interceptor.
+    ///
+    /// Provided version string must be ASCII.
+    pub fn with_accept_metadata(mut self, version: &str) -> Result<Self, anyhow::Error> {
+        let accept_value = format!("application/vnd.miden; version={version}");
+        self.metadata.insert("accept", AsciiMetadataValue::try_from(accept_value)?);
+        Ok(self)
+    }
+}
+
+impl Interceptor for MetadataInterceptor {
+    fn call(&mut self, request: Request<()>) -> Result<Request<()>, Status> {
+        let mut request = request;
+        for (key, value) in &self.metadata {
+            request.metadata_mut().insert(*key, value.clone());
+        }
+        Ok(request)
+    }
+}
+
+// TYPE ALIASES FOR INSTRUMENTED CLIENTS
+// ================================================================================================
+
+pub type RpcClient =
+    generated::rpc::api_client::ApiClient<InterceptedService<Channel, MetadataInterceptor>>;
+pub type BlockProducerClient =
+    generated::block_producer::api_client::ApiClient<InterceptedService<Channel, OtelInterceptor>>;
+pub type StoreNtxBuilderClient = generated::ntx_builder_store::ntx_builder_client::NtxBuilderClient<
+    InterceptedService<Channel, OtelInterceptor>,
+>;
+pub type StoreBlockProducerClient =
+    generated::block_producer_store::block_producer_client::BlockProducerClient<
+        InterceptedService<Channel, OtelInterceptor>,
+    >;
+pub type StoreRpcClient =
+    generated::rpc_store::rpc_client::RpcClient<InterceptedService<Channel, OtelInterceptor>>;
+
+// BUILDER CONFIGURATION
+// ================================================================================================
+
+/// Builder for configuring and creating gRPC clients with consistent options.
+///
+/// This builder provides a fluent interface for configuring various aspects of gRPC client
+/// connections including TLS, timeouts, metadata, and connection types. It supports both
+/// eager and lazy connection establishment.
+///
+/// # Examples
+///
+/// ```rust,no_run
+/// use miden_node_proto::clients::{Builder, Rpc, RpcClient};
+/// use std::time::Duration;
+///
+/// # async fn example() -> anyhow::Result<()> {
+/// // Create a client with TLS and timeout
+/// let client: RpcClient = Builder::new()
+///     .with_address("https://rpc.example.com:8080".to_string())
+///     .with_tls()
+///     .with_timeout(Duration::from_secs(30))
+///     .connect::<Rpc>()
+///     .await?;
+/// # Ok(())
+/// # }
+/// ```
+#[derive(Default, Clone)]
+pub struct Builder {
+    /// Whether to enable TLS encryption for the connection.
+    pub with_tls: bool,
+    /// The gRPC server address in the format `{protocol}://{hostname}:{port}`.
+    pub address: String,
+    /// Optional timeout for gRPC operations.
+    pub with_timeout: Option<Duration>,
+    /// Optional version string to include in request metadata.
+    pub metadata_version: Option<String>,
```

Comment:
Similar question as above - does this need to handle the `genesis` part of the accept header somehow as well?